tests/mock_crust/tunnel.rs

```diff
@@ -147,3 +147,74 @@ fn tunnel_clients() {
                                          min_section_size);
     verify_invariant_for_all_nodes(&nodes);
 }
+
+#[test]
+fn tunnel_peer_connect_failure() {
+    let min_section_size = 5;
+    let network = Network::new(min_section_size, None);
+    network.block_connection(Endpoint(2), Endpoint(3));
+    network.block_connection(Endpoint(3), Endpoint(2));
+    let mut nodes = create_connected_nodes(&network, min_section_size);
+
+    network.send_crust_event(Endpoint(2), crust::Event::ConnectFailure(PeerId(3)));
+    let _ = poll_all(&mut nodes, &mut []);
+    verify_invariant_for_all_nodes(&nodes);
+}
+
+// Verifies tunnel switched.
+fn verify_tunnel_switch(nodes: &mut Vec<TestNode>, node: usize, client_1: usize, client_2: usize) {
+    while let Ok(event) = nodes[client_1].inner.try_next_ev() {
+        match event {
+            Event::NodeLost(name, _) => {
+                assert!(name == nodes[node].name() || name == nodes[client_2].name());
+            }
+            Event::NodeAdded(name, _) => assert!(name == nodes[client_2].name()),
+            _ => {
+                panic!("{:?} received unexpected event {:?}",
+                       nodes[client_1].name(),
+                       event)
+            }
+        }
+    }
+    while let Ok(event) = nodes[client_2].inner.try_next_ev() {
+        match event {
+            Event::NodeLost(name, _) => assert!(name == nodes[client_1].name()),
+            Event::NodeAdded(name, _) => assert!(name == nodes[client_1].name()),
+            _ => {
+                panic!("{:?} received unexpected event {:?}",
+                       nodes[client_2].name(),
+                       event)
+            }
+        }
+    }
+}
+
+#[test]
+fn tunnel_node_disrupted() {
+    let min_section_size = 5;
+    let network = Network::new(min_section_size, None);
+    network.block_connection(Endpoint(2), Endpoint(3));
+    network.block_connection(Endpoint(3), Endpoint(2));
+    let mut nodes = create_connected_nodes(&network, min_section_size);
+
+    network.send_crust_event(Endpoint(2), crust::Event::LostPeer(PeerId(0)));
+    network.send_crust_event(Endpoint(0), crust::Event::LostPeer(PeerId(2)));
+    poll_and_resend(&mut nodes, &mut []);
+    verify_tunnel_switch(&mut nodes, 0, 2, 3);
```

Comment:
I don't really like the style of this test. Can't we instead add a hook to let us check that firstly, 2 and 3 connect via 0, and secondly that they reconnect (it might be enough just to check the invariant at the end, rather than how the nodes are connected)?